python/benchmark/gen_data_distributed.py

```diff
@@ -0,0 +1,554 @@
+#
+# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import argparse
+import random
+import sys
+from abc import abstractmethod
+from typing import Any, Dict, Iterator, List, Optional, Tuple, Union
+
+import numpy as np
+import pandas as pd
+from pyspark.mllib.random import RandomRDDs
+from pyspark.sql import DataFrame, SparkSession
+from pyspark.sql.functions import array
+from sklearn.datasets import (
+    make_blobs,
+    make_classification,
+    make_low_rank_matrix,
+    make_regression,
+)
+
+from benchmark.utils import WithSparkSession, inspect_default_params_from_func, to_bool
+
+
+def dtype_to_pyspark_type(dtype: Union[np.dtype, str]) -> str:
+    """Convert np.dtype to the corresponding pyspark type"""
+    dtype = np.dtype(dtype)
+    if dtype == np.float32:
+        return "float"
+    elif dtype == np.float64:
+        return "double"
+    else:
+        raise RuntimeError("Unsupported dtype, found ", dtype)
+
+
+class DataGen(object):
+    """DataGen interface"""
+
+    @abstractmethod
+    def gen_dataframe(self, spark: SparkSession) -> Tuple[DataFrame, List[str]]:
+        raise NotImplementedError()
+
+
+class DataGenBase(DataGen):
```

Comment:
Same here.  Prefer to import common functions and classes from gen_data.py. If file gen_data.py will be deleted, try to create a new file (e.g. gen_data_utils.py) to host common functions and classes.

Reply:
OK - I'll import common functions from gen_data for now, and once we've tested and benchmarked both implementations, I can move the definitions over into the gen_data_distributed file. Does that work?

Reply:
Also, I've added the --test argument option in DataGenBase, which wasn't originally in gen_data. OK if I redefine the class in gen_data_distributed?

Reply:
> OK - I'll import common functions from gen_data for now, and once we've tested and benchmarked both implementations, I can move the definitions over into the gen_data_distributed file. Does that work?

Work for me. Merging can be done later.

Reply:
> Also, I've added the --test argument option in DataGenBase, which wasn't originally in gen_data. OK if I redefine the class in gen_data_distributed?

Good question. You may redefine by extending the class to create a subclass (e.g. DistributedDataGenBase). 

I feel a better way is to avoid '--test' because it seems an argument used in the test code and it has no effect on user code (i.e. python benchmark/gen_data_distributed.py). You can avoid '--test' by adding an additional variable (e.g. Optional[Dict[str, Any]]) to return values of function DataGenBase::gen_dataframe(..) to return meta information(e.g. centers of BlobsDataGen).

Reply:
With this optional return value, I would still need some way of knowing whether to return the optional argument at all. Wouldn't I still need some flag to indicate that the method call is for testing?